pyo3-macros-backend/src/pyfunction.rs

```diff
@@ -83,13 +84,105 @@ impl PyFunctionArgPyO3Attributes {
     }
 }
 
+type DeprecatedFunctionMessageAttribute = KeywordAttribute<attributes::kw::message, LitStr>;
```

Comment:
```suggestion
type DeprecatedFunctionMessageAttribute = KeywordAttribute<attributes::kw::message, LitCStr>;
```
Since we know the string literal at compile time, and `PyErr_WarnEx` takes a null terminated string, we can avoid making the CString in `warn_bound` . We don't have an api for raising warnings that way, so you need to make a function for that as well.

Reply:
I used `PyErr::warn_bound` and that seems to do the conversion to CString and matches what you refer to as as the new API :)

Reply:
The reference is here https://pyo3.rs/main/doc/src/pyo3/err/mod.rs#605-619

Reply:
> I used `PyErr::warn_bound` and that seems to do the conversion to CString and matches what you refer to as the new API :)

Right, but what I mean is that we can move this from runtime to compile time. See https://github.com/PyO3/pyo3/pull/4255 for example.

We could have a function like this:
```rust
   fn warn_with_cstr<'py>(
        py: Python<'py>,
        category: &Bound<'py, PyAny>,
        message: &CStr,
        stacklevel: i32,
    ) -> PyResult<()> {
        error_on_minusone(py, unsafe {
            ffi::PyErr_WarnEx(
                category.as_ptr(),
                message.as_ptr(),
                stacklevel as ffi::Py_ssize_t,
            )
        })
    }
```

Reply:
Thanks for the clarification! Moving that to compile time makes a lot sense :)

Reply:
I don't think we can parse it as `LitCStr` because of MSRV. But we can make use of the `utils::LitCStr` compatibility struct that we introduced lately to to conditionally emit either a c-string literal or `ffi::c_str!`.

Reply:
That makes sense. I just saw the MSRV check. Will work on that later. :)